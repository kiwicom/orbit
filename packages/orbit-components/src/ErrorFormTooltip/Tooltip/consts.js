// @flow

export const SIDE_NUDGE = 15;
