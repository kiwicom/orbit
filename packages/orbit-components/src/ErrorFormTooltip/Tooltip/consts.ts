export const SIDE_NUDGE = 10;
