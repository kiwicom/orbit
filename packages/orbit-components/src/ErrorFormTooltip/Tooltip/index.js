// @flow
import * as React from "react";
import styled, { css } from "styled-components";

import useClickOutside from "../../hooks/useClickOutside";
import KEY_CODE_MAP from "../../common/keyMaps";
import handleKeyDown from "../../utils/handleKeyDown";
import defaultTheme from "../../defaultTheme";
import media from "../../utils/mediaQuery";
import { StyledText } from "../../Text";
import { Item } from "../../List/ListItem";
import CloseIc from "../../icons/Close";
import { rtlSpacing, right } from "../../utils/rtl";
import resolveColor from "./helpers/resolveColor";
import tooltipArrowStyle from "./helpers/tooltipArrowStyle";
import resolveTooltipArrowPosition from "./helpers/resolveTooltipArrowPosition";
import resolveTooltipPosition from "./helpers/resolveTooltipPosition";
import useDimensions from "../hooks/useDimensions";
import { POSITIONS, SIDE_NUDGE } from "./consts";

import type { Props } from ".";

const StyledFormFeedbackTooltip = styled.div`
  ${({ theme, isHelp, shown, inputSize }) => css`
    display: flex;
    justify-content: space-between;
    align-items: center;
    position: absolute;
    box-sizing: border-box;
    border-radius: ${theme.orbit.borderRadiusNormal};
    box-shadow: ${theme.orbit.boxShadowElevatedLevel1};
    padding: ${inputSize === "small" ? theme.orbit.spaceXSmall : theme.orbit.spaceSmall};
    padding-${right}: ${isHelp && theme.orbit.spaceSmall};
    z-index: 10;
    max-height: none;
    overflow: visible;
    width: ${`calc(100% + ${SIDE_NUDGE * 2}px)`};
    background-color: ${resolveColor};
    visibility: ${shown ? "visible" : "hidden"};
    opacity: ${shown ? "1" : "0"};
    transition: opacity ${theme.orbit.durationFast} ease-in-out,
      visibility ${theme.orbit.durationFast} ease-in-out;

    /* prevent position, IEs don't have initial YAY */
    top: auto;
    right: auto;
    bottom: auto;
    left: auto;

    img {
      max-width: 100%;
    }

    &::after {
      width: 0;
      height: 0;
      border-style: solid;
      content: " ";
      display: block;
      position: absolute;

      ${tooltipArrowStyle};
      ${resolveTooltipArrowPosition};
    }

    ${media.largeMobile(css`
      width: auto;
    `)};

    ${resolveTooltipPosition};
`}
`;

// $FlowFixMe: https://github.com/flow-typed/flow-typed/issues/3653#issuecomment-568539198
StyledFormFeedbackTooltip.defaultProps = {
  theme: defaultTheme,
};

const StyledTooltipContent = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    height: 100%;
    font-family: ${theme.orbit.fontFamily};
    font-size: ${theme.orbit.fontSizeTextNormal};
    font-weight: ${theme.orbit.fontWeightNormal};
    line-height: ${theme.orbit.lineHeightText};
    color: ${theme.orbit.paletteWhite};

    & ${StyledText}, ${Item}, a {
      color: ${theme.orbit.paletteWhite};
      font-weight: ${theme.orbit.fontWeightNormal};
    }

    ${media.largeMobile(css`
      font-size: ${theme.orbit.fontSizeTextSmall};
      font-weight: ${theme.orbit.fontWeightMedium};

      ${StyledText}, ${Item}, a {
        font-weight: ${theme.orbit.fontWeightMedium};
        font-size: ${theme.orbit.fontSizeTextSmall};
      }
      a:hover {
        color: ${theme.orbit.paletteWhite};
      }
      a:focus {
        color: ${theme.orbit.paletteWhite};
      }
    `)};
  `}
`;

// $FlowFixMe: https://github.com/flow-typed/flow-typed/issues/3653#issuecomment-568539198
StyledTooltipContent.defaultProps = {
  theme: defaultTheme,
};

const StyledCloseButton = styled.a`
  ${({ theme }) => css`
    color: ${theme.orbit.paletteWhite};
    cursor: pointer;
    margin: ${rtlSpacing(`0 0 0 ${theme.orbit.spaceSmall}`)};
    display: flex;
  `}
`;

// $FlowFixMe: https://github.com/flow-typed/flow-typed/issues/3653#issuecomment-568539198
StyledCloseButton.defaultProps = {
  theme: defaultTheme,
};

const ErrorFormTooltip = ({
  onShown,
  labelRef,
  dataTest,
  inputSize,
  iconRef,
  children,
  shown,
  isHelp = false,
  inlineLabel,
  id,
}: Props): React.Node => {
  const contentRef = React.useRef(null);
  const tooltipRef = React.useRef<HTMLElement | null>(null);

  const dimensions = useDimensions({
    labelRef,
    iconRef,
    contentRef: tooltipRef,
    children,
    inlineLabel,
  });

  const preferredPosition = React.useMemo(
    () => (dimensions.label.top - dimensions.content.height > 0 ? POSITIONS.TOP : POSITIONS.BOTTOM),
    [dimensions.label.top, dimensions.content.height],
  );

  useClickOutside(contentRef, () => {
    if (!isHelp) onShown(false);
  });

  React.useEffect(() => {
    // $FlowFixMe: TODO
    const link = tooltipRef.current?.querySelector("a");

    const handleTab = ev => {
      if (isHelp) return;
      if (ev.keyCode === KEY_CODE_MAP.TAB && link) {
        onShown(true);
        if (document.activeElement === link) {
          onShown(false);
        }
      } else {
        onShown(false);
      }
    };

    window.addEventListener("keydown", handleTab);
    return () => {
      window.removeEventListener("keydown", handleTab);
    };
  }, [onShown, isHelp]);

  return (
    <StyledFormFeedbackTooltip
      ref={tooltipRef}
      inputSize={inputSize}
      position={preferredPosition}
      shown={shown && dimensions.set}
      isHelp={isHelp}
      data-test={dataTest}
      inlineLabel={inlineLabel}
      aria-live="polite"
      id={id}
      {...dimensions}
    >
      <StyledTooltipContent ref={contentRef}>{children}</StyledTooltipContent>
      {isHelp && (
        <StyledCloseButton
          tabIndex={0}
          // $FlowFixMe: TODO
          onKeyDown={handleKeyDown(onShown)}
          onClick={ev => {
            ev.preventDefault();
            if (shown) onShown(false);
          }}
        >
          <CloseIc ariaLabel="close" size="small" />
        </StyledCloseButton>
      )}
    </StyledFormFeedbackTooltip>
  );
};

export default ErrorFormTooltip;
