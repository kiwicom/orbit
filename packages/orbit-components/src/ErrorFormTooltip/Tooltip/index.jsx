// @flow
import * as React from "react";
import styled, { css } from "styled-components";
import { usePopper } from "react-popper";

import useClickOutside from "../../hooks/useClickOutside";
import KEY_CODE_MAP from "../../common/keyMaps";
import handleKeyDown from "../../utils/handleKeyDown";
import defaultTheme from "../../defaultTheme";
import media from "../../utils/mediaQuery";
import { StyledText } from "../../Text";
import { Item } from "../../List/ListItem";
import CloseIc from "../../icons/Close";
import { rtlSpacing, right } from "../../utils/rtl";
import resolveColor from "./helpers/resolveColor";
import resolvePlacement from "./helpers/resolvePlacement";
import { SIDE_NUDGE } from "./consts";
import useTheme from "../../hooks/useTheme";

import type { Props } from ".";

const StyledArrow = styled.div`
  ${({ position, top, left, bottom, right: popperRight, transform }) => css`
    position: ${position};
    top: ${top};
    left: ${left};
    right: ${popperRight};
    bottom: ${bottom};
    transform: ${transform};
    &:before {
      content: "";
      background: ${resolveColor};
      width: 0.6rem;
      height: 0.6rem;
      transform: translate(-50%, -50%) rotate(45deg);
      position: absolute;
    }
  `}
`;

const StyledFormFeedbackTooltip = styled.div`
  ${({
    theme,
    isHelp,
    shown,
    inputSize,
    top,
    left,
    position,
    bottom,
    right: popperRight,
    transform,
  }) => css`
    display: flex;
    justify-content: space-between;
    box-sizing: border-box;
    border-radius: ${theme.orbit.borderRadiusNormal};
    box-shadow: ${theme.orbit.boxShadowElevatedLevel1};
    padding: ${theme.orbit.spaceXSmall} ${
    inputSize === "small" ? theme.orbit.spaceXSmall : theme.orbit.spaceSmall
  };
    padding-${right}: ${isHelp && theme.orbit.spaceSmall};
    z-index: 10;
    max-height: none;
    overflow: visible;
    width: ${`calc(100% + ${SIDE_NUDGE * 2}px)`};
    width: min(${`calc(100% + ${SIDE_NUDGE * 2}px)`}, 100vw);
    background: ${resolveColor};
    visibility: ${shown ? "visible" : "hidden"};
    opacity: ${shown ? "1" : "0"};
    transition: opacity ${theme.orbit.durationFast} ease-in-out,
      visibility ${theme.orbit.durationFast} ease-in-out;

    position: ${position};
    top: ${top};
    left: ${left};
    bottom: ${bottom};
    right: ${popperRight};
    transform: ${transform};

    img {
      max-width: 100%;
    }

    ${StyledArrow} {
      ${resolvePlacement};
    }

    ${media.largeMobile(css`
      width: auto;
    `)};
`}
`;

// $FlowFixMe: https://github.com/flow-typed/flow-typed/issues/3653#issuecomment-568539198
StyledArrow.defaultProps = {
  theme: defaultTheme,
};

// $FlowFixMe: https://github.com/flow-typed/flow-typed/issues/3653#issuecomment-568539198
StyledFormFeedbackTooltip.defaultProps = {
  theme: defaultTheme,
};

const StyledTooltipContent = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    height: 100%;
    font-family: ${theme.orbit.fontFamily};
    font-size: ${theme.orbit.fontSizeTextNormal};
    font-weight: ${theme.orbit.fontWeightNormal};
    line-height: ${theme.orbit.lineHeightText};
    color: ${theme.orbit.paletteWhite};

    & ${StyledText}, ${Item}, a {
      color: ${theme.orbit.paletteWhite};
      font-size: ${theme.orbit.fontSizeTextNormal};
      font-weight: ${theme.orbit.fontWeightNormal};
      &:hover,
      &:focus {
        color: ${theme.orbit.paletteWhite};
      }
    }

    ${media.largeMobile(css`
      font-size: ${theme.orbit.fontSizeTextSmall};
      font-weight: ${theme.orbit.fontWeightMedium};

      ${StyledText}, ${Item}, a {
        font-weight: ${theme.orbit.fontWeightMedium};
        font-size: ${theme.orbit.fontSizeTextSmall};
      }
    `)};
  `}
`;

// $FlowFixMe: https://github.com/flow-typed/flow-typed/issues/3653#issuecomment-568539198
StyledTooltipContent.defaultProps = {
  theme: defaultTheme,
};

const StyledCloseButton = styled.a`
  ${({ theme }) => css`
    color: ${theme.orbit.paletteWhite};
    cursor: pointer;
    margin: ${rtlSpacing(`0 0 0 ${theme.orbit.spaceSmall}`)};
    display: flex;
  `}
`;

// $FlowFixMe: https://github.com/flow-typed/flow-typed/issues/3653#issuecomment-568539198
StyledCloseButton.defaultProps = {
  theme: defaultTheme,
};

const ErrorFormTooltip = ({
  onShown,
  dataTest,
  helpClosable,
  inputSize,
  children,
  shown,
  referenceElement,
  inlineLabel,
  isHelp = false,
  id,
}: Props): React.Node => {
  const contentRef = React.useRef(null);
  const { rtl } = useTheme();
  const tooltipRef = React.useRef(null);
  const [arrowRef, setArrowRef] = React.useState(null);

  const resolveOffset = React.useCallback(() => {
    if (inlineLabel) {
      if (inputSize === "small") return [rtl ? 10 : -14, 7];
      return [rtl ? 6 : -6, 6];
    }
    return [rtl ? SIDE_NUDGE : -SIDE_NUDGE, 7];
  }, [inlineLabel, inputSize, rtl]);

  const { styles, attributes: attrs, update } = usePopper(
    referenceElement?.current,
    tooltipRef.current,
    {
      placement: rtl ? "top-end" : "top-start",
      modifiers: [
        {
          name: "offset",
          options: {
            offset: resolveOffset,
          },
        },
        {
          name: "arrow",
          options: {
            element: arrowRef,
          },
        },
        {
          name: "eventListeners",
          options: {
            scroll: false,
          },
        },
      ],
    },
  );

  const { popper, arrow } = styles;

  useClickOutside(tooltipRef, () => {
    onShown(false);
  });

  React.useEffect(() => {
    if (update) update();
  }, [update, resolveOffset, shown]);

  React.useEffect(() => {
    const link = tooltipRef.current?.querySelector("a");
    const handleTab = ev => {
      if (isHelp) return;
      if (ev.keyCode === KEY_CODE_MAP.TAB && link) {
        onShown(true);
        if (document.activeElement === link) {
          onShown(false);
        }
      } else {
        onShown(false);
      }
    };

    window.addEventListener("keydown", handleTab);
    return () => {
      window.removeEventListener("keydown", handleTab);
    };
  }, [onShown, isHelp, helpClosable]);

  return (
    <StyledFormFeedbackTooltip
      id={id}
      ref={tooltipRef}
      inputSize={inputSize}
      shown={shown}
      isHelp={isHelp}
      data-test={dataTest}
      aria-live="polite"
      inlineLabel={inlineLabel}
      placement={attrs.popper && attrs.popper["data-popper-placement"]}
      position={popper.position}
      top={popper.top}
      left={popper.left}
      right={popper.right}
      bottom={popper.bottom}
      transform={popper.transform}
    >
      <StyledArrow
        isHelp={isHelp}
        ref={setArrowRef}
        position={popper.position}
        placement={attrs.popper && attrs.popper["data-popper-placement"]}
        top={arrow.top}
        left={arrow.left}
        right={arrow.right}
        bottom={arrow.bottom}
      />
      <StyledTooltipContent ref={contentRef}>{children}</StyledTooltipContent>
      {isHelp && helpClosable && (
        <StyledCloseButton
          tabIndex={0}
          // $FlowFixMe: TODO
          onKeyDown={handleKeyDown(onShown)}
          onClick={ev => {
            ev.preventDefault();
            if (shown) onShown(false);
          }}
        >
          <CloseIc ariaLabel="close" size="small" />
        </StyledCloseButton>
      )}
    </StyledFormFeedbackTooltip>
  );
};

export default ErrorFormTooltip;
