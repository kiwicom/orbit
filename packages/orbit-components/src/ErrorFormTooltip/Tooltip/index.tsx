import * as React from "react";
import styled, { css } from "styled-components";
import { usePopper } from "react-popper";

import * as Common from "../../common/types";
import useClickOutside from "../../hooks/useClickOutside";
import KEY_CODE_MAP from "../../common/keyMaps";
import handleKeyDown from "../../utils/handleKeyDown";
import defaultTheme from "../../defaultTheme";
import media from "../../utils/mediaQuery";
import { StyledText } from "../../Text";
import { Item } from "../../List/ListItem";
import CloseIc from "../../icons/Close";
import { rtlSpacing, right } from "../../utils/rtl";
import resolveColor from "./helpers/resolveColor";
import resolvePlacement from "./helpers/resolvePlacement";
import { SIDE_NUDGE } from "./consts";
import useTheme from "../../hooks/useTheme";
import { Props } from "./types";

const StyledArrow = styled.div<{
  inlineLabel?: boolean;
  isHelp?: boolean;
  inputSize?: Common.InputSize;
  placement?: string;
}>`
  position: absolute;
  ${resolvePlacement};
  &:before {
    content: "";
    background: ${resolveColor};
    width: 8px;
    height: 8px;
    transform: translate(-50%, -50%) rotate(45deg);
    position: absolute;
  }
`;

const StyledFormFeedbackTooltip = styled.div<{
  isHelp?: boolean;
  shown?: boolean;
  inlineLabel?: boolean;
  inputSize?: Common.InputSize;
  placement?: string;
  top?: string | number;
  left?: string | number;
  position?: string | number;
  bottom?: string | number;
  right?: string | number;
  transform?: string;
}>`
  ${({
    theme,
    isHelp,
    shown,
    inputSize,
    top,
    left,
    position,
    bottom,
    right: popperRight,
    transform,
  }) => css`
    display: flex;
    justify-content: space-between;
    box-sizing: border-box;
    border-radius: ${theme.orbit.borderRadiusNormal};
    box-shadow: ${theme.orbit.boxShadowElevatedLevel1};
    padding: ${theme.orbit.spaceXSmall} ${
    inputSize === "small" ? theme.orbit.spaceXSmall : theme.orbit.spaceSmall
  };
    padding-${right}: ${isHelp && theme.orbit.spaceSmall};
    z-index: 10;
    max-height: none;
    overflow: visible;
    width: min(${`calc(100% - ${SIDE_NUDGE * 2}px)`}, 100vw);
    background: ${resolveColor};
    visibility: ${shown ? "visible" : "hidden"};
    opacity: ${shown ? "1" : "0"};
    transition: opacity ${theme.orbit.durationFast} ease-in-out,
      visibility ${theme.orbit.durationFast} ease-in-out;

    position: ${position};
    top: ${top};
    left: ${left};
    bottom: ${bottom};
    right: ${popperRight};
    transform: ${transform};

    img {
      max-width: 100%;
    }

    ${media.largeMobile(css`
      width: auto;
    `)};
`}
`;

StyledArrow.defaultProps = {
  theme: defaultTheme,
};

StyledFormFeedbackTooltip.defaultProps = {
  theme: defaultTheme,
};

const StyledTooltipContent = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    height: 100%;
    font-family: ${theme.orbit.fontFamily};
    font-size: ${theme.orbit.fontSizeTextNormal};
    font-weight: ${theme.orbit.fontWeightNormal};
    line-height: ${theme.orbit.lineHeightText};
    color: ${theme.orbit.paletteWhite};

    & ${StyledText}, ${Item}, a {
      color: ${theme.orbit.paletteWhite};
      font-size: ${theme.orbit.fontSizeTextNormal};
      font-weight: ${theme.orbit.fontWeightNormal};
      &:hover,
      &:focus {
        color: ${theme.orbit.paletteWhite};
      }
    }

    ${media.largeMobile(css`
      font-size: ${theme.orbit.fontSizeTextSmall};
      font-weight: ${theme.orbit.fontWeightMedium};

      ${StyledText}, ${Item}, a {
        font-weight: ${theme.orbit.fontWeightMedium};
        font-size: ${theme.orbit.fontSizeTextSmall};
      }
    `)};
  `}
`;

StyledTooltipContent.defaultProps = {
  theme: defaultTheme,
};

const StyledCloseButton = styled.a`
  ${({ theme }) => css`
    color: ${theme.orbit.paletteWhite};
    cursor: pointer;
    margin: ${rtlSpacing(`0 0 0 ${theme.orbit.spaceSmall}`)};
    display: flex;
  `}
`;

StyledCloseButton.defaultProps = {
  theme: defaultTheme,
};

const ErrorFormTooltip = ({
  onShown,
  dataTest,
  helpClosable,
  inputSize,
  children,
  shown,
  referenceElement,
  inlineLabel,
  isHelp = false,
  id,
}: Props) => {
  const contentRef = React.useRef<HTMLDivElement | null>(null);
  const tooltipRef = React.useRef<HTMLDivElement | null>(null);
  const [arrowRef, setArrowRef] = React.useState<HTMLDivElement | null>(null);
  const { rtl } = useTheme();

  const resolveOffset = React.useCallback(() => {
    if (inlineLabel) {
      if (inputSize === "small") return [0, 4];
    }
    return [0, 3];
  }, [inlineLabel, inputSize]);

  const { styles, attributes: attrs, update } = usePopper(
    referenceElement?.current,
    tooltipRef.current,
    {
      placement: rtl ? "top-end" : "top-start",
      modifiers: [
        {
          name: "offset",
          options: {
            // @ts-expect-error TODO
            offset: resolveOffset,
          },
        },
        {
          name: "flip",
          enabled: false,
        },
        {
          name: "arrow",
          options: {
            element: arrowRef,
          },
        },
        {
          name: "eventListeners",
          options: {
            scroll: false,
          },
        },
      ],
    },
  );

  const { popper } = styles;

  useClickOutside(tooltipRef, () => {
    onShown(false);
  });

  React.useEffect(() => {
    if (update) update();
  }, [update, resolveOffset, shown]);

  React.useEffect(() => {
    const link = tooltipRef.current?.querySelector("a");
    const handleTab = ev => {
      if (isHelp) return;
      if (ev.keyCode === KEY_CODE_MAP.TAB && link) {
        onShown(true);
        if (document.activeElement === link) {
          onShown(false);
        }
      } else {
        onShown(false);
      }
    };

    window.addEventListener("keydown", handleTab);
    return () => {
      window.removeEventListener("keydown", handleTab);
    };
  }, [onShown, isHelp, helpClosable]);

  return (
    <StyledFormFeedbackTooltip
      id={id}
      ref={tooltipRef}
      inputSize={inputSize}
      shown={shown}
      isHelp={isHelp}
      data-test={dataTest}
      aria-live="polite"
      inlineLabel={inlineLabel}
      placement={attrs.popper && attrs.popper["data-popper-placement"]}
      position={popper.position}
      top={popper.top}
      left={popper.left}
      right={popper.right}
      bottom={popper.bottom}
      transform={popper.transform}
    >
      <StyledArrow
        isHelp={isHelp}
        ref={setArrowRef}
        inputSize={inputSize}
        inlineLabel={inlineLabel}
        placement={attrs.popper && attrs.popper["data-popper-placement"]}
      />
      <StyledTooltipContent ref={contentRef}>{children}</StyledTooltipContent>
      {isHelp && helpClosable && (
        <StyledCloseButton
          tabIndex={0}
          // @ts-expect-error TODO
          onKeyDown={handleKeyDown<HTMLAnchorElement>(onShown)}
          onClick={ev => {
            ev.preventDefault();
            if (shown) onShown(false);
          }}
        >
          <CloseIc ariaLabel="close" size="small" />
        </StyledCloseButton>
      )}
    </StyledFormFeedbackTooltip>
  );
};

export default ErrorFormTooltip;
