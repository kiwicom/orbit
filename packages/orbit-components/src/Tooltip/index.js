// @flow
import * as React from "react";

import { SIZE_OPTIONS } from "./consts";
import TooltipPrimitive from "../primitives/TooltipPrimitive";
import useMediaQuery from "../hooks/useMediaQuery";
import MobileDialog from "../primitives/MobileDialogPrimitive";

import type { Props } from "./index";

const Tooltip = ({
  children,
  enabled = true,
  tabIndex = "0",
  dataTest,
  size = SIZE_OPTIONS.SMALL,
  content,
  preferredPosition,
  preferredAlign,
  stopPropagation = false,
  removeUnderlinedText,
}: Props) => {
  const { isLargeMobile } = useMediaQuery();
  return isLargeMobile ? (
    <TooltipPrimitive
      dataTest={dataTest}
      tabIndex={tabIndex}
      enabled={enabled}
      content={content}
      size={size}
      preferredPosition={preferredPosition}
      preferredAlign={preferredAlign}
      stopPropagation={stopPropagation}
      removeUnderlinedText={removeUnderlinedText}
    >
      {children}
    </TooltipPrimitive>
  ) : (
    <MobileDialog
      dataTest={dataTest}
      tabIndex={tabIndex}
      enabled={enabled}
      content={content}
      removeUnderlinedText={removeUnderlinedText}
      stopPropagation={stopPropagation}
    >
      {children}
    </MobileDialog>
  );
};

export default Tooltip;
