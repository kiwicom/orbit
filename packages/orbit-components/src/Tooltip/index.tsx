import * as React from "react";

import { SIZE_OPTIONS } from "./consts";
import TooltipPrimitive from "../primitives/TooltipPrimitive";
import useMediaQuery from "../hooks/useMediaQuery";
import MobileDialog from "../primitives/MobileDialogPrimitive";
import { Props } from "./types";

const Tooltip = ({
  children,
  enabled = true,
  tabIndex = "0",
  dataTest,
  size = SIZE_OPTIONS.SMALL,
  content,
  id,
  placement,
  lockScrolling,
  renderInPortal = true,
  stopPropagation = false,
  removeUnderlinedText,
  block = false,
}: Props) => {
  const { isLargeMobile } = useMediaQuery();
  return isLargeMobile ? (
    <TooltipPrimitive
      dataTest={dataTest}
      id={id}
      tabIndex={tabIndex}
      enabled={enabled}
      content={content}
      size={size}
      renderInPortal={renderInPortal}
      placement={placement}
      stopPropagation={stopPropagation}
      removeUnderlinedText={removeUnderlinedText}
      block={block}
    >
      {children}
    </TooltipPrimitive>
  ) : (
    <MobileDialog
      dataTest={dataTest}
      tabIndex={tabIndex}
      id={id}
      enabled={enabled}
      lockScrolling={lockScrolling}
      content={content}
      removeUnderlinedText={removeUnderlinedText}
      stopPropagation={stopPropagation}
      block={block}
    >
      {children}
    </MobileDialog>
  );
};

export default Tooltip;
