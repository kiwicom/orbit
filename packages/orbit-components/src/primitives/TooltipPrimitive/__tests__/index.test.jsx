// @flow
import * as React from "react";
import { render, screen, act } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import Tooltip from "..";
import Airplane from "../../../icons/Airplane";

describe("Tooltip", () => {
  it("it should match snapshot", async () => {
    const content = "Write some message to the user";
    const { container } = render(
      <Tooltip content={content}>
        <Airplane />
      </Tooltip>,
    );

    expect(container.firstChild).toMatchSnapshot();
    // Needs to flush async `floating-ui` hooks
    // https://github.com/floating-ui/floating-ui/issues/1520
    // $FlowFixMe
    await act(async () => {});
  });

  it("should call onClick 1 time", async () => {
    const content = "Write some message to the user";
    const onClick = jest.fn();

    render(
      <div onClick={onClick} role="button" tabIndex={0} onKeyDown={() => {}}>
        <Tooltip
          stopPropagation
          content={
            <div onClick={onClick} role="button" tabIndex={0} onKeyDown={() => {}}>
              {content}
            </div>
          }
        >
          <div>kek</div>
        </Tooltip>
      </div>,
    );

    userEvent.hover(screen.getByText("kek"));
    userEvent.click(screen.getByText(content));

    expect(onClick).toHaveBeenCalledTimes(1);
    // Needs to flush async `floating-ui` hooks
    // https://github.com/floating-ui/floating-ui/issues/1520
    // $FlowFixMe
    await act(async () => {});
  });

  it("should call onClick 2 times", async () => {
    const content = "Write some message to the user";
    const onClick = jest.fn();

    render(
      <div onClick={onClick} role="button" tabIndex={0} onKeyDown={() => {}}>
        <Tooltip
          content={
            <div onClick={onClick} role="button" tabIndex={0} onKeyDown={() => {}}>
              {content}
            </div>
          }
        >
          <div>kek</div>
        </Tooltip>
      </div>,
    );

    userEvent.hover(screen.getByText("kek"));
    userEvent.click(screen.getByText(content));

    expect(onClick).toHaveBeenCalledTimes(2);
    // Needs to flush async `floating-ui` hooks
    // https://github.com/floating-ui/floating-ui/issues/1520
    // $FlowFixMe
    await act(async () => {});
  });
});
