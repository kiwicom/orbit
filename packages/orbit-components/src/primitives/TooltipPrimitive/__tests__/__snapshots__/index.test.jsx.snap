// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip it should match snapshot 1`] = `
.c1 {
  display: inline-block;
  width: 24px;
  height: 24px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  vertical-align: middle;
  fill: currentColor;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c0:focus:active {
  outline: none;
}

.c0 [disabled] {
  pointer-events: none;
}

<span
  aria-describedby="1"
  class="c0"
  tabindex="0"
>
  <svg
    class="c1"
    preserveAspectRatio="xMidYMid meet"
    viewBox="0 0 24 24"
  >
    <path
      d="M7.85 7.331a469.495 469.495 0 0 1 2.358 2.219l.041.04c.105.109.237.245.412.252.197.008 3.043-.034 3.137-.046h.003c.094-.013.243-.032.165-.204-.079-.172-3.788-2.696-4.3-2.996-.073-.043-.135-.082-.19-.116-.323-.2-.39-.241-1.124-.211l-.213.008c-.669.026-.786.03-.887.229-.09.174.093.346.373.61.07.065.145.137.224.215Zm10.41 5.919c1.637.007 2.41-.766 2.41-1.35 0-.585-.228-1.193-1.85-1.412-1.082-.146-5.197-.136-12.345.032-.151-.004-.184-.016-.311-.12L5.045 9.258c-.235-.24-.711-.364-1.16-.357-.449.006-.596.25-.56.469.036.22.083.37.14.51l1.166 2.886c.144.323.425.52.796.484l4.408-.023c.144 0 .204.08.204.168-.002.04-.026.116-.15.248a635.858 635.858 0 0 0-2.713 2.913c-.244.261-.563.605-.537.857.029.286.39.314.63.314.29 0 .464 0 .608.002l.796-.002h.016c.671-.004.804-.005 1.146-.314.254-.23 1.743-1.595 4.468-4.094a.613.613 0 0 1 .293-.066c1.336.007 2.557.006 3.665-.003Z"
    />
  </svg>
</span>
`;
